libtorch-ffi/src/Torch/Internal/Unmanaged/Optim.hs

```diff
@@ -0,0 +1,284 @@
+    for(int i=0;i<init_params->size();i++){
+      params.push_back((*init_params)[i].detach().set_requires_grad(true));
+    }
+    auto options = torch::optim::SGDOptions($(double lr))
+      .momentum($(double momentum))
+      .dampening($(double dampening))
+      .weight_decay($(double weight_decay))
+      .nesterov($(bool nesterov));
+    torch::optim::SGD* optimizer = new torch::optim::SGD(params, options);
+    optimizer->zero_grad();
+    return dynamic_cast<torch::optim::Optimizer*>(optimizer);
+  }|]
+
+adam
+  :: CDouble
+  -> CDouble
+  -> CDouble
+  -> CDouble
+  -> CDouble
+  -> CBool
+  -> Ptr TensorList
+  -> IO (Ptr Optimizer)
+adam adamLr adamBetas0 adamBetas1 adamEps adamWeightDecay adamAmsgrad initParams =
+  [C.throwBlock| torch::optim::Optimizer* {
+    std::vector<at::Tensor>* init_params = $(std::vector<at::Tensor>* initParams);
+    std::vector<at::Tensor> params;
+    for(int i=0;i<init_params->size();i++){
+      params.push_back((*init_params)[i].detach().set_requires_grad(true));
+    }
+    auto options = torch::optim::AdamOptions()
+      .lr($(double adamLr))
+      .betas(std::make_tuple($(double adamBetas0),$(double adamBetas1)))
+      .eps($(double adamEps))
+      .weight_decay($(double adamWeightDecay))
+      .amsgrad($(bool adamAmsgrad));
+    torch::optim::Adam* optimizer = new torch::optim::Adam(params, options);
+    optimizer->zero_grad();
+    return dynamic_cast<torch::optim::Optimizer*>(optimizer);
+  }|]
+
+adamw
+  :: CDouble
+  -> CDouble
+  -> CDouble
+  -> CDouble
+  -> CDouble
+  -> CBool
+  -> Ptr TensorList
+  -> IO (Ptr Optimizer)
+adamw adamLr adamBetas0 adamBetas1 adamEps adamWeightDecay adamAmsgrad initParams =
+  [C.throwBlock| torch::optim::Optimizer* {
+    std::vector<at::Tensor>* init_params = $(std::vector<at::Tensor>* initParams);
+    std::vector<at::Tensor> params;
+    for(int i=0;i<init_params->size();i++){
+      params.push_back((*init_params)[i].detach().set_requires_grad(true));
+    }
+    auto options = torch::optim::AdamWOptions()
+      .lr($(double adamLr))
+      .betas(std::make_tuple($(double adamBetas0),$(double adamBetas1)))
+      .eps($(double adamEps))
+      .weight_decay($(double adamWeightDecay))
+      .amsgrad($(bool adamAmsgrad));
+    torch::optim::AdamW* optimizer = new torch::optim::AdamW(params, options);
+    optimizer->zero_grad();
+    return dynamic_cast<torch::optim::Optimizer*>(optimizer);
+  }|]
+
+
+lbfgs
+  :: CDouble
+  -> CInt
+  -> CInt
+  -> CDouble
+  -> CDouble
+  -> CInt
+  -> Maybe (Ptr StdString)
+  -> Ptr TensorList
+  -> IO (Ptr Optimizer)
+lbfgs lr max_iter max_eval tolerance_grad tolerance_change history_size Nothing initParams =
+  [C.throwBlock| torch::optim::Optimizer* {
+    std::vector<at::Tensor>* init_params = $(std::vector<at::Tensor>* initParams);
+    std::vector<at::Tensor> params;
+    for(int i=0;i<init_params->size();i++){
+      params.push_back((*init_params)[i].detach().set_requires_grad(true));
```

Comment:
still confused. these `params`, are they model or optimizer parameters?

Reply:
These `params` are model's one not optimizer's one.